compartmentalize-grains-via-pop.md

```diff
@@ -0,0 +1,71 @@
+provided by salt with POP.  We propose that this project, codenamed `corn` replaces grains in salt while
+maintaining backwards compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+"The original implementation of grains was not designed to scale to the level that salt requires" - thatch45
+Tom wrote POP to solve this problem and we're going to use it. 
+
+By using Plugin Oriented Programming, grains can be separated into individual projects that can be maintained
+by the community safely.
+
+This also solves the issue of `core.py` being an unmaintainable monolith.  It allows platform specific grains
+to be independently maintained.
+
+Working group leaders and other community members focused on specific platforms can manage the release cycle for
+individual grains projects independently from salt.
+
+# Design
+[design]: #detailed-design
+
+Grains are currently created by calling functions in the salt loader.  We would simply need to call
+external POP functions to collect grains.  We need to maintain backwards compatibility with custom user grains.
+
+- All salt grains have already been ported to and tested on the new platform.
+- The new platform has many more features than the salt-grains implementation, but is compatible with salt grains.
+- None of the grain names have been changed, but the output of grains is more consistent/complete between platforms.
+- This will resolve many bugs that are open against grains and will migrate those bugs OUT of the salt
+platform -- making salt easier to develop
+- Grains in the new platform are independently and fully tested.
+- Most grains tests will be able to be removed from salt; speeding up it's test suite.
+- This comprises a complete rewrite of ALL grains.  Backwards compatibility is expected, but we need this to get
+in early so that any unintentional issues can be resolved quickly.
+- Salt releases will be able to pin idem-grain dependencies with each release.
+- Users will be to update their grains implementation to a custom idem-grain version.
+
+## Alternatives
+[alternatives]: #alternatives
+
+Keep things the way they are.  Does this mean we would suffer in stagnation?  Will we be stuck maintaining a monolith.
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+What will further adoption of POP mean inside of salt?  The purpose of this sep and these features is not to provide
+a definitive answer to this question -- rather to introduce plugin oriented programming into salt in a very safe way.
+In the future this will determine the design so that salt can be further compartmentalized and written in POP.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this? Please consider:
+
+- We will be introducing a number of new dependencies.
+- The new implementation of grains/pop is written with python3.6 and later with asyncio
+(The latest release of salt is also python3.6 and later).
```

Comment:
Which release is this?
We currently support 3.5, will Magnesium be >= 3.6?

Reply:
It would have to be >=python3.6 to support pop code, this is something we should discuss with the team in greater detail.  @dmurphy18 What is tying us to python 3.5?

Reply:
@Akm0d Debian 9, Ubuntu 16.04, SLES are the last supported platforms still using Py 3.5

Reply:
@dmurphy18 I handled this in the code.  Grainsv2 won't be included in python3.5 and lower builds.  The `__hub__` will also not be available on <=python3.5

Reply:
I'm definitely not a big fan of having another dunder added to salt. It seems like sprawl in lack of a properly architected design change.